_posts/2023-02-22-user-defined-operators.md

```diff
@@ -0,0 +1,484 @@
+effects extend to wherever the type is visible, not just the current source unit.
+
+#### More Examples
+
+##### Arithmetic Operations and Comparisons
+In this example we define multiple operators:
+
+```solidity
+pragma solidity ^0.8.19;
+
+type Int is int;
+
+function add(Int a, Int b) pure returns (Int) {
+    return Int.wrap(Int.unwrap(a) + Int.unwrap(b));
+}
+
+function sub(Int a, Int b) pure returns (Int) {
+    return Int.wrap(Int.unwrap(a) + Int.unwrap(b));
+}
+
+function unsub(Int a) pure returns (Int) {
+    return Int.wrap(-Int.unwrap(a));
+}
+
+function equals(Int a, Int b) pure returns (bool) {
+    return Int.unwrap(a) == Int.unwrap(b);
+}
+
+using {add as +, sub as -, unsub as -} for Int global;
+using {equals as ==} for Int global;
+
+function test(Int a, Int b) pure returns (Int) {
+    return a == b ? a + b : -a - b;
+}
+```
+
+##### Unchecked Counter
+This example shows how operators can be used to bypass the checked arithmetic for types
+specifically defined for cases where those checks are superfluous:
+
+```solidity
+pragma solidity ^0.8.19;
+
+type UncheckedCounter is uint;
+
+using {add as +, lt as <} for UncheckedCounter global;
+
+UncheckedCounter constant ONE = UncheckedCounter.wrap(1);
+
+function add(UncheckedCounter x, UncheckedCounter y) pure returns (UncheckedCounter) {
+    unchecked {
+        return UncheckedCounter.wrap(
+            UncheckedCounter.unwrap(x) +
+            UncheckedCounter.unwrap(y)
+        );
+    }
+}
+
+function lt(UncheckedCounter x, UncheckedCounter y) pure returns (bool) {
+    return UncheckedCounter.unwrap(x) < UncheckedCounter.unwrap(y);
+}
+
+contract C {
+    uint internalCounter = 12;
+
+    function testCounter() public returns (uint) {
+        for (
+            UncheckedCounter i = UncheckedCounter.wrap(12);
+            i < UncheckedCounter.wrap(20);
+            i = i + ONE
+        )
+            ++internalCounter;
+
+        return internalCounter;
+    }
+}
+```
+
+### AST Changes
+The new feature introduces two new attributes to the Abstract Syntax Tree (AST) produced by the compiler:
+1. `functionList` on `UsingForDirective` nodes can now contain operator entries of the following structure:
+    ```json
+    {
+        "operator": "+",
+        "definition": { /* function */ }
+    }
+    ```
+    in addition to the current entries for functions that are structured as follows:
+    ```json
+    {
+        "function": { /* function */ }
+    }
+    ```
+1. `BinaryOperation` and `UnaryOperation` nodes may now have a `function` attribute.
+    If present, the attribute indicates that the operation is actually an invocation of a user-defined
+    operator and specifies the ID of the function used to define it.
+
+This effectively introduces a new way to perform a function call, not represented by a `FunctionCall` node.
+This will affect any tool that needs to detect and follow function calls, for example, to build a
+control flow graph.
```

Comment:
```suggestion
This effectively introduces a new way to perform a function call not represented by a `FunctionCall` node.
This will affect any tool that needs to detect and follow function calls, such as building a
control flow graph.
```

Reply:
It does not sound correct to me. After the change the second sentence does not really connect with the first one, it refers to a different subject. Not sure how to reword it to make this suggestion work - the way I'd do it is really how I did it already.